quic/s2n-quic-tls/src/callback.rs

```diff
@@ -213,6 +224,12 @@ where
                     _ => {
                         let (key, header_key) =
                             OneRttKey::new(self.endpoint, aead_algo, pair).expect("invalid cipher");
+                        unsafe {
```

Comment:
Okay I am willing to be really wrong about this, but please explain if I am: doesn't the client move from the initial space to the handshake space _after_ reading the server hello? The client can only calculate handshake keys after it's read the server hello, so I think this can be in the HandshakePhase::Initial space.

Reply:
So i believe you are correct and the limitation currently is our unit test + polling setup.

- the poll method does alot of things including sending and receiving [link](https://github.com/aws/s2n-quic/blob/main/quic/s2n-quic-rustls/src/session.rs#L170) (this makes it difficult to call only receive and only transmit in isolation).
- therefore we poll client, poll server and then check progress. [link](https://github.com/aws/s2n-quic/blob/main/quic/s2n-quic-core/src/crypto/tls/testing.rs#L139-L153)

I believe if we could only receive and called receive on the client, then we would be able to test alpn available after the handshake_keys.

```
---- client_server_test stdout ----
Client: s2n_quic_rustls::session::Session: rx initial
Client: s2n_quic_rustls::session::Session: tx initial
Client: s2n_quic_rustls::session::Session: rx initial
Server: s2n_quic_rustls::session::Session: rx initial
Server: s2n_quic_rustls::session::Session: rx initial
1/2 RTT
------------ 1
Client: s2n_quic_rustls::session::Session: rx initial
Client: s2n_quic_rustls::session::Session: rx initial
Server: s2n_quic_rustls::session::Session: rx initial
Server: s2n_quic_rustls::session::Session: tx initial
Server: s2n_quic_rustls::session::Session: server name
Server: s2n_quic_rustls::session::Session: application protocol
Server: s2n_quic_rustls::session::Session: handshake keys
Server: s2n_quic_rustls::session::Session: tx handshake        <--------------- server should send server_hello here
Server: s2n_quic_rustls::session::Session: 1-rtt keys
Server: s2n_quic_rustls::session::Session: rx handshake
1/2 RTT
------------ 2
Client: s2n_quic_rustls::session::Session: rx initial
Client: s2n_quic_rustls::session::Session: tx initial
Client: s2n_quic_rustls::session::Session: handshake keys
Client: s2n_quic_rustls::session::Session: rx handshake           <-------- note the rx handshake only in iteration 2
Client: s2n_quic_rustls::session::Session: tx handshake
Client: s2n_quic_rustls::session::Session: application protocol
Client: s2n_quic_rustls::session::Session: 1-rtt keys
Client: s2n_quic_rustls::session::Session: rx handshake
Client: s2n_quic_rustls::session::Session: handshake complete
Server: s2n_quic_rustls::session::Session: rx handshake
Server: s2n_quic_rustls::session::Session: rx handshake
1/2 RTT
------------ 3
Client: s2n_quic_rustls::session::Session: rx application
Server: s2n_quic_rustls::session::Session: rx handshake
Server: s2n_quic_rustls::session::Session: handshake complete
...
```

Reply:
Here is a [patch](https://github.com/aws/s2n-quic/pull/1238/commits/85b59d94a3f5e77f2c9663de3bbe6f9c48d69e50) which passes tests. it shows that even though we have generated handshake keys, we havnt actually read form the handshake rx buffer (havnt read the server hello yet). 

This is something we can improve upon (s2n-tls with its callbacks might be more difficult), but I dont want to tackle that work as part of this PR so I am going to revert and add todo/issue to capture that work.

Reply:
I think your question of how do we generate handshake keys when we havnt read the ServerHello is still valid and I am not sure I know how to answer that.

Is the ServerHello being split, or some other kind of magic? This is still an open question.

Reply:
opened an issue to track the work: https://github.com/aws/s2n-quic/issues/1240